src/visual.ts

```diff
@@ -1088,52 +1340,111 @@ export class BulletChart implements IVisual {
+                .attr("font-size", PixelConverter.fromPoint(model.settings.axis.unitsFont.fontSize.value))
+                .attr("font-weight", model.settings.axis.unitsFont.bold.value ? "bold" : "normal")
+                .attr("font-style", model.settings.axis.unitsFont.italic.value ? "italic" : "normal")
+                .attr("text-decoration", model.settings.axis.unitsFont.underline.value ? "underline" : "none")
                 .text(measureUnitsText);
         }
 
-        if (this.interactivityService) {
-            const targetCollection = this.data.barRects.concat(this.data.valueRects);
-            const behaviorOptions: BulletBehaviorOptions = {
-                rects: bullets,
-                groupedRects: groupedBulletsSelection,
-                valueRects: valueSelection,
-                clearCatcher: this.clearCatcher,
-                interactivityService: this.interactivityService,
-                bulletChartSettings: this.data.settings,
-                hasHighlights: this.data.hasHighlights,
-                behavior: this.behavior,
-                dataPoints: targetCollection
-            };
-
-            this.interactivityService.bind(behaviorOptions);
+        const targetCollection = this.data.barRects.concat(this.data.valueRects);
+        const behaviorOptions: BehaviorOptions = {
+            dataPoints: targetCollection,
+            hasHighlights: this.data.hasHighlights,
+            rects: bullets,
+            groupedRects: groupedBulletsSelection,
+            valueRects: valueSelection,
+            clearCatcher: this.bulletBody,
+        };
+
+        this.behavior.bindEvents(behaviorOptions);
+
+        this.tooltipServiceWrapper.addTooltip(valueSelection, (data: BarRect) => data.tooltipInfo);
+        this.tooltipServiceWrapper.addTooltip(bullets, (data: BarRect) => data.tooltipInfo);
+    }
+
+    private renderLegend(definedColors: DefinedColors): void {
+        const legendDataObject: LegendData = {
+            dataPoints: [],
+            title: this.visualSettings.legend.titleText.value,
+            fontSize: this.visualSettings.legend.font.fontSize.value,
+            fontFamily: this.visualSettings.legend.font.fontFamily.value,
+            labelColor: this.visualSettings.legend.labelColor.value.value,
+        };
+
+        const colors = this.visualSettings.colors.getData();
+        const filtered: { displayNameKey: string; color: string }[] = [];
+
+        if (definedColors) {
+            Object.entries(definedColors).forEach(([key, value]) => {
+                if (value) {
+                    const color: { displayNameKey: string; color: string; } = colors[key];
+                    filtered.push(color);
+                }
+            });
+        } else {
+            filtered.push(
+                ...Object.values(colors).map((color) => ({ displayNameKey: color.displayNameKey, color: color.color }))
```

Comment:
it looks like we don't need to `map()` here. we are returning the same object spending resources for nothing. 
If I'm wrong seems it would be nice to add TS features here.

Reply:
That's not the case. The colors looks like this:
```TypeScript
const colors = {
    minColor: { displayNameKey: this.minColor.displayNameKey, color: this.minColor.value.value },
    needsImprovementColor: { displayNameKey: this.needsImprovementColor.displayNameKey, color: this.needsImprovementColor.value.value },
    satisfactoryColor: { displayNameKey: this.satisfactoryColor.displayNameKey, color: this.satisfactoryColor.value.value },
    goodColor: { displayNameKey: this.goodColor.displayNameKey, color: this.goodColor.value.value },
    veryGoodColor: { displayNameKey: this.veryGoodColor.displayNameKey, color: this.veryGoodColor.value.value },
    bulletColor: { displayNameKey: this.bulletColor.displayNameKey, color: this.bulletColor.value.value },
};
```
I'm using a key like `minColor` to get the rigth displayName/color if `definedColors` object is not undefined.
Otherwise, I'll add all colors to legend. The line you selected converts an object into an array.